### PR TITLE
implement h2d transfers using pinned memory

### PR DESCRIPTION
**Pull request by robik75**
# What ❔

This PR implements h2d transfers using pinned memory buffer.

## Why ❔

Pinned memory transfers are faster and can execute asynchronously.

## Checklist

- [x] PR title corresponds to the body of PR (we generate changelog entries from PRs).
- [x] Tests for the changes have been added / updated.
- [x] Documentation comments have been added / updated.
- [x] Code has been formatted via `cargo fmt` and linted via `cargo check`.

### CODE REVIEW

File: src/test.rs [hunk 1]
```diff
@@ -328,7 +328,13 @@ fn clone_reference_tree(
...
 #[ignore]
 fn test_dry_runs() {
-    let (setup_cs, finalization_hint) = init_or_synth_cs_for_sha256::<DevCSConfig, true>(None);
+    let (setup_cs, finalization_hint) =
+        init_or_synth_cs_for_sha256::<DevCSConfig, Global, true>(None);
+    let (proving_cs, _) =
+        init_or_synth_cs_for_sha256::<ProvingCSConfig, Global, true>(finalization_hint.as_ref());
+    let witness = proving_cs.witness.unwrap();
+    let (reusable_cs, _) =
+        init_or_synth_cs_for_sha256::<ProvingCSConfig, Global, false>(finalization_hint.as_ref());
```
Comment (mcarilli) [inline src/test.rs:337]: does this just correct something Dima missed (adding the Global allocator generic parameter)?
If so, why did you move them?

Comment (robik75) [reply src/test.rs:337]: Changes were needed because of dima´s changes to boojum´s API. The move was not intended but should not make a difference.

File: src/primitives/mem.rs [hunk 2]
```diff
@@ -9,6 +16,72 @@ pub fn h2d<T>(host: &[T], device: &mut [T]) -> CudaResult<()> {
...
+        let mut pending_callbacks = vec![];
+        for (i, (src, dst)) in host
+            .chunks(chunk_size)
+            .zip(device.chunks_mut(chunk_size))
+            .enumerate()
+        {
+            let idx = i % STREAMS_COUNT;
+            let stream = &streams[idx];
+            let buffer_offset = idx * chunk_size;
+            let buffer = &buffer[buffer_offset..buffer_offset + src.len()];
```
Comment (mcarilli) [inline src/primitives/mem.rs:65]: is overwriting `buffer` here intentional?

Comment (robik75) [reply src/primitives/mem.rs:65]: yes

File: src/primitives/mem.rs [hunk 3]
```diff
@@ -9,6 +16,72 @@ pub fn h2d<T>(host: &[T], device: &mut [T]) -> CudaResult<()> {
...
+                            dst_chunk.as_mut_ptr(),
+                            src_chunk.len(),
+                        )
+                    })
+                }
+            });
+        };
+        events[0].record(main_stream)?;
+        for stream in streams.iter() {
+            stream.wait_event(&events[0], CudaStreamWaitEventFlags::DEFAULT)?;
```
[the remark this answers is not in the document]
Comment (robik75) [reply src/primitives/mem.rs:54]: not only those, but really anything happening on the main stream, otherwise the aux streams just would run ahead on their own regardless of whether the main stream is already there

**Review by mcarilli — DISMISSED** [review]

**Review by mcarilli — APPROVED** [review]
